doc/optimgaika.1
.\" Copyright (C) 2011 Kyrill Detinov <lazy.kent@opensuse.org>
.\"
.\" This manual page is distributed under the terms
.\" of the GNU Free Documentation License version 1.3.
.\"
.TH TRIMAGE "1" "2011-07-11" "optimgaika 1.0.5" "User Commands"

.SH NAME
optimgaika \- losslessly optimizing png and jpeg liles

.SH SYNOPSIS
.B optimgaika
.RI [ options ]

.SH DESCRIPTION
Front\-end to compress png and jpeg images via optipng, advpng, pngcrush
and jpegoptim.

.SH OPTIONS
.TP
\fB\-d\fI directory\fR, \fB\-\-directory\fR=\fIdirectory\fR
Compresses images in directory.
.TP
\fB\-f\fI filename\fR, \fB\-\-file\fR=\fIfilename\fR
Compresses image.
.TP
\fB\-h\fR, \fB\-\-help\fR
Show help message.
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Quiet mode.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Verbose mode (default).
.TP
\fB\-\-version\fR
Show program version number.

.SH "SEE ALSO"
.BR advpng (1),
.BR jpegoptim (1),
.BR opt-png (1),
.BR opt-jpg (1),
.BR pngcrush (1),
.BR pngrecolor (1),
.BR pngstrip (1)


debian/changelog
optimgaika (1.2) xenial; urgency=medium

  * The original author of trimage decided to not merge my pull request (https://github.com/Kilian/Optimgaika/pull/44), so I am forking Optimgaika and naming the fork as optimgaika (оптимгайка)
  * This is the first version after renaming the fork
  * Bumping version from 1.1 to 1.2

 -- Mikhail Novosyolov <mikhailnov@dumalogiya.ru>  Sat, 30 Sep 2017 15:20:00 +0300
 
trimage (1.1) xenial; urgency=medium

  * change jpegoptim to guetzli
  * Add Russian translation in trimage.desktop

 -- Mikhail Novosyolov <mikhailnov@dumalogiya.ru>  Sun, 09 Jul 2017 13:05:00 +0300
 
trimage (1.0.5-0ubuntu1) jaunty; urgency=low

  * prevent images from becoming larger after recompression
  * prevent images from being written when not writeable
  * allow dragging of entire directories
  * performance increases for large numbers of files

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Sun, 26 Sep 2010 15:24:04 +0200

trimage (1.0.4-0ubuntu1) jaunty; urgency=low

  * prevent systemtray from starting if a CLI compression is done
  * preliminary OSX work

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Fri, 17 Sep 2010 20:45:16 +0200

trimage (1.0.3-0ubuntu1) jaunty; urgency=low

  * fix for environments without supported system tray
  * add copyright info for filesize.py
  * update various debian configurations

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Sat, 12 Jun 2010 14:04:35 +0200

trimage (1.0.2-0ubuntu1) jaunty; urgency=low

  * save window geometry and file directory upon closing
  * prettier systray menu
  * Switch to dpkg-source 3.0(quilt) format

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Thu, 03 Jun 2010 13:01:30 +0200

trimage (1.0.1b2-0ubuntu1) jaunty; urgency=low

  * use a threadpool for images
  * more robust file handling
  * re-adding images now results in recompressing them
  * compressing message now shows filename
  * wider array of status messages in the table
  * add a notification area icon
  * remove BOM
 -- Kilian Valkhof <kilian@kdesktop>   Mon, 29 Mar 2010 13:08:18 +0200

trimage (1.0.1b-0ubuntu1) jaunty; urgency=low

  * use a threadpool for images
  * more robust file handling
  * re-adding images now results in recompressing them
  * compressing message now shows filename
  * wider array of status messages in the table
  * add a notification area icon
 -- Kilian Valkhof <kilian@kdesktop>   Mon, 29 Mar 2010 13:08:18 +0200

trimage (1.0.0b3-0ubuntu1) jaunty; urgency=low

  * better image filetype determination
  * use unicode strings everywhere
  * fix bug where Optimgaika would crash if a directory didn't end in a slash
  * avoid a segfault when parsing empty directories
  * ../../debian/changelog
 -- Kilian Valkhof <kilian@kdesktop>   Mon, 29 Mar 2010 13:08:18 +0200

trimage (1.0.0b2-0ubuntu1) jaunty; urgency=low

  * correct parsing of unicode characters in filenames
  * changelog
 -- Kilian Valkhof <kilian@kdesktop>   Sun, 28 Mar 2010 16:09:45 +0200

trimage (1.0.0b-0ubuntu1) jaunty; urgency=low

  * Optimgaika image compressor
 -- Kilian Valkhof <help@trimage.org>   Tue, 23 Mar 2010 20:18:17 +0100



src/optimgaika/optimgaika.py
#!/usr/bin/python
import time
import sys
import errno
from os import listdir
from os import path
from os import remove
from os import access
from os import W_OK as WRITEABLE
from shutil import copy
from subprocess import call, PIPE
from optparse import OptionParser

from PyQt4.QtCore import *
from PyQt4.QtGui import *
from filesize import *
from imghdr import what as determinetype

from Queue import Queue
from ThreadPool import ThreadPool
from multiprocessing import cpu_count

from ui import Ui_optimgaika

# change version to 1.1 after replacing jpegoptim with guetzli
VERSION = "1.1"


class StartQT4(QMainWindow):

    def __init__(self, parent=None):
        QWidget.__init__(self, parent)
        self.ui = Ui_optimgaika()
        self.ui.setupUi(self)

        self.showapp = True
        self.verbose = True
        self.imagelist = []

        QCoreApplication.setOrganizationName("Kilian Valkhof")
        QCoreApplication.setOrganizationDomain("optimgaika.org")
        QCoreApplication.setApplicationName("Optimgaika")
        self.settings = QSettings()
        self.restoreGeometry(self.settings.value("geometry").toByteArray())

        # check if apps are installed
        if self.checkapps():
            quit()

        #add quit shortcut
        if hasattr(QKeySequence, "Quit"):
            self.quit_shortcut = QShortcut(QKeySequence(QKeySequence.Quit),
                self)
        else:
            self.quit_shortcut = QShortcut(QKeySequence("Ctrl+Q"), self)

        # disable recompress
        self.ui.recompress.setEnabled(False)
        #self.ui.recompress.hide()

        # make a worker thread
        self.thread = Worker()

        # connect signals with slots
        QObject.connect(self.ui.addfiles, SIGNAL("clicked()"),
            self.file_dialog)
        QObject.connect(self.ui.recompress, SIGNAL("clicked()"),
            self.recompress_files)
        QObject.connect(self.quit_shortcut, SIGNAL("activated()"),
            qApp, SLOT('quit()'))
        QObject.connect(self.ui.processedfiles, SIGNAL("fileDropEvent"),
            self.file_drop)
        QObject.connect(self.thread, SIGNAL("finished()"), self.update_table)
        QObject.connect(self.thread, SIGNAL("terminated()"), self.update_table)
        QObject.connect(self.thread, SIGNAL("updateUi"), self.update_table)

        self.compressing_icon = QIcon(QPixmap(self.ui.get_image("pixmaps/compressing.gif")))

        # activate command line options
        self.commandline_options()

        if QSystemTrayIcon.isSystemTrayAvailable() and not self.cli:
            self.systemtray = Systray(self)

    def commandline_options(self):
        self.cli = False
        """Set up the command line options."""
        parser = OptionParser(version="%prog " + VERSION,
            description="GUI front-end to compress png and jpg images via "
                "optipng, advpng and guetzli")

        parser.set_defaults(verbose=True)
        parser.add_option("-v", "--verbose", action="store_true",
            dest="verbose", help="Verbose mode (default)")
        parser.add_option("-q", "--quiet", action="store_false",
            dest="verbose", help="Quiet mode")

        parser.add_option("-f", "--file", action="store", type="string",
            dest="filename", help="compresses image and exit")
        parser.add_option("-d", "--directory", action="store", type="string",
            dest="directory", help="compresses images in directory and exit")

        options, args = parser.parse_args()

        # make sure we quit after processing finished if using cli
        if options.filename or options.directory:
            QObject.connect(self.thread, SIGNAL("finished()"), quit)
            self.cli = True

        # send to correct function
        if options.filename:
            self.file_from_cmd(options.filename.decode("utf-8"))
        if options.directory:
            self.dir_from_cmd(options.directory.decode("utf-8"))

        self.verbose = options.verbose

    """
    Input functions
    """

    def dir_from_cmd(self, directory):
        """
        Read the files in the directory and send all files to compress_file.
        """
        self.showapp = False
        dirpath = path.abspath(directory)
        imagedir = listdir(directory)
        filelist = [path.join(dirpath, image) for image in imagedir]
        self.delegator(filelist)

    def file_from_cmd(self, image):
        """Get the file and send it to compress_file"""
        self.showapp = False
        filelist = [path.abspath(image)]
        self.delegator(filelist)

    def file_drop(self, images):
        """
        Get a file from the drag and drop handler and send it to compress_file.
        """
        self.delegator(images)

    def file_dialog(self):
        """Open a file dialog and send the selected images to compress_file."""
        fd = QFileDialog(self)
        fd.restoreState(self.settings.value("fdstate").toByteArray())
        directory = self.settings.value("directory", QVariant("")).toString()
        fd.setDirectory(directory)

        images = fd.getOpenFileNames(self,
            "Select one or more image files to compress",
            directory,
            # this is a fix for file dialog differentiating between cases
            "Image files (*.png *.jpg *.jpeg *.PNG *.JPG *.JPEG)")

        self.settings.setValue("fdstate", QVariant(fd.saveState()))
        if images:
          self.settings.setValue("directory", QVariant(path.dirname(unicode(images[0]))))
        self.delegator([unicode(fullpath) for fullpath in images])

    def recompress_files(self):
        """Send each file in the current file list to compress_file again."""
        self.delegator([row.image.fullpath for row in self.imagelist])

    """
    Compress functions
    """

    def delegator(self, images):
        """
        Receive all images, check them and send them to the worker thread.
        """
        delegatorlist = []
        for fullpath in images:
            try: # recompress images already in the list
                image = (i.image for i in self.imagelist
                    if i.image.fullpath == fullpath).next()
                if image.compressed:
                    image.reset()
                    image.recompression = True
                    delegatorlist.append(image)
            except StopIteration:
            	if not path.isdir(fullpath):
                    self. add_image(fullpath, delegatorlist)
                else:
                    self.walk(fullpath, delegatorlist)

        self.update_table()
        self.thread.compress_file(delegatorlist, self.showapp, self.verbose,
            self.imagelist)

    def walk(self, dir, delegatorlist):
        """
        Walks a directory, and executes a callback on each file
        """
        dir = path.abspath(dir)
        for file in [file for file in listdir(dir) if not file in [".","..",".svn",".git",".hg",".bzr",".cvs"]]:
            nfile = path.join(dir, file)

            if path.isdir(nfile):
                self.walk(nfile, delegatorlist)
            else:
                self.add_image(nfile, delegatorlist)

    def add_image(self, fullpath, delegatorlist):
        """
        Adds an image file to the delegator list and update the tray and the title of the window
        """
        image = Image(fullpath)
        if image.valid:
            delegatorlist.append(image)
            self.imagelist.append(ImageRow(image, self.compressing_icon))
            if QSystemTrayIcon.isSystemTrayAvailable() and not self.cli:
                self.systemtray.trayIcon.setToolTip("Optimgaika image compressor (" + str(len(self.imagelist)) + " files)")
                self.setWindowTitle("Optimgaika image compressor (" + str(len(self.imagelist)) + " files)")
        else:
            print >> sys.stderr, u"[error] %s not a supported image file and/or not writeable" % image.fullpath

    """
    UI Functions
    """

    def update_table(self):
        """Update the table view with the latest file data."""
        tview = self.ui.processedfiles
        # set table model
        tmodel = TriTableModel(self, self.imagelist,
            ["Filename", "Old Size", "New Size", "Compressed"])
        tview.setModel(tmodel)

        # set minimum size of table
        vh = tview.verticalHeader()
        vh.setVisible(False)

        # set horizontal header properties
        hh = tview.horizontalHeader()
        hh.setStretchLastSection(True)

        # set all row heights
        nrows = len(self.imagelist)
        for row in range(nrows):
            tview.setRowHeight(row, 25)

        # set the second column to be longest
        tview.setColumnWidth(0, 300)

        # enable recompress button
        self.enable_recompress()

    """
    Helper functions
    """

    def enable_recompress(self):
        """Enable the recompress button."""
        self.ui.recompress.setEnabled(True)
        if QSystemTrayIcon.isSystemTrayAvailable() and not self.cli:
            self.systemtray.recompress.setEnabled(True)

    def checkapps(self):
        """Check if the required command line apps exist."""
        exe = ".exe" if (sys.platform == "win32") else ""
        status = False
        
        # guetzli does the compression job much better than jpegoptim (and much slower)
        #retcode = self.safe_call("jpegoptim" + exe + " --version")
        #if retcode != 0:
            #status = True
            #sys.stderr.write("[error] please install jpegoptim")

        retcode = self.safe_call("optipng" + exe + " -v")
        if retcode != 0:
            status = True
            sys.stderr.write("[error] please install optipng")

        retcode = self.safe_call("advpng" + exe + " --version")
        if retcode != 0:
            status = True
            sys.stderr.write("[error] please install advancecomp")

        retcode = self.safe_call("pngcrush" + exe + " -version")
        if retcode != 0:
            status = True
            sys.stderr.write("[error] please install pngcrush")
        return status
        
        # guetzli currently does not have neither --version not --help flags
        retcode = self.safe_call("guetzli" + exe)
        if retcode != 0:
            status = True
            sys.stderr.write("[error] please install guetzli")
        return status

    def safe_call(self, command):
        """ cross-platform command-line check """
        while True:
            try:
                return call(command, shell=True, stdout=PIPE)
            except OSError, e:
                if e.errno == errno.EINTR:
                    continue
                else:
                    raise

    def hide_main_window(self):
        if self.isVisible():
            self.hide()
            if QSystemTrayIcon.isSystemTrayAvailable():
                self.systemtray.hideMain.setText("&Show window")
        else:
            self.show()
            if QSystemTrayIcon.isSystemTrayAvailable():
                self.systemtray.hideMain.setText("&Hide window")

    def closeEvent(self, event):
      self.settings.setValue("geometry", QVariant(self.saveGeometry()))
      event.accept()


class TriTableModel(QAbstractTableModel):

    def __init__(self, parent, imagelist, header, *args):
        """
        @param parent Qt parent object.
        @param imagelist A list of tuples.
        @param header A list of strings.
        """
        QAbstractTableModel.__init__(self, parent, *args)
        self.imagelist = imagelist
        self.header = header

    def rowCount(self, parent):
        """Count the number of rows."""
        return len(self.imagelist)

    def columnCount(self, parent):
        """Count the number of columns."""
        return len(self.header)

    def data(self, index, role):
        """Fill the table with data."""
        if not index.isValid():
            return QVariant()
        elif role == Qt.DisplayRole:
            data = self.imagelist[index.row()][index.column()]
            return QVariant(data)
        elif index.column() == 0 and role == Qt.DecorationRole:
            # decorate column 0 with an icon of the image itself
            f_icon = self.imagelist[index.row()][4]
            return QVariant(f_icon)
        else:
            return QVariant()

    def headerData(self, col, orientation, role):
        """Fill the table headers."""
        if orientation == Qt.Horizontal and (role == Qt.DisplayRole or
        role == Qt.DecorationRole):
            return QVariant(self.header[col])
        return QVariant()


class ImageRow:

    def __init__(self, image, waitingIcon=None):
        """ Build the information visible in the table image row. """
        self.image = image
        d = {
            'shortname': lambda i: self.statusStr() % i.shortname,
            'oldfilesizestr': lambda i: size(i.oldfilesize, system=alternative)
                if i.compressed else "",
            'newfilesizestr': lambda i: size(i.newfilesize, system=alternative)
                if i.compressed else "",
            'ratiostr': lambda i:
                "%.1f%%" % (100 - (float(i.newfilesize) / i.oldfilesize * 100))
                if i.compressed else "",
            'icon': lambda i: i.icon if i.compressed else waitingIcon,
            'fullpath': lambda i: i.fullpath, #only used by cli
        }
        names = ['shortname', 'oldfilesizestr', 'newfilesizestr',
                      'ratiostr', 'icon']
        for i, n in enumerate(names):
            d[i] = d[n]

        self.d = d

    def statusStr(self):
        """ Set the status message. """
        if self.image.failed:
            return "ERROR: %s"
        if self.image.compressing:
            message = "Compressing %s..."
            return message
        if not self.image.compressed and self.image.recompression:
            return "Queued for recompression..."
        if not self.image.compressed:
            return "Queued..."
        return "%s"

    def __getitem__(self, key):
        return self.d[key](self.image)


class Image:

    def __init__(self, fullpath):
        """ gather image information. """
        self.valid = False
        self.reset()
        self.fullpath = fullpath
        if path.isfile(self.fullpath) and access(self.fullpath, WRITEABLE):
            self.filetype = determinetype(self.fullpath)
            if self.filetype in ["jpeg", "png"]:
                oldfile = QFileInfo(self.fullpath)
                self.shortname = oldfile.fileName()
                self.oldfilesize = oldfile.size()
                self.icon = QIcon(self.fullpath)
                self.valid = True

    #def _determinetype(self):
    #    """ Determine the filetype of the file using imghdr. """
    #    filetype = determinetype(self.fullpath)
    #    if filetype in ["jpeg", "png"]:
    #        self.filetype = filetype
    #    else:
    #        self.filetype = None
    #    return self.filetype

    def reset(self):
        self.failed = False
        self.compressed = False
        self.compressing = False
        self.recompression = False

    def compress(self):
        """ Compress the image and return it to the thread. """
        if not self.valid:
            raise "Tried to compress invalid image (unsupported format or not \
            file)"
        self.reset()
        self.compressing = True
        exe = ".exe" if (sys.platform == "win32") else ""
        runString = {
            # "jpeg": u"jpegoptim" + exe + " -f --strip-all '%(file)s'",
            # I commented out jpegoptim because guetzli compresses MUCH BETTER having a similar output quality
            # the default quality level for guetzli, if the parameter '--quality Q' is not passed, it 95.
            "jpeg": u"guetzli" + exe + " --quality 89 '%(file)s' '%(file)s.bak' && mv '%(file)s.bak' '%(file)s'",
            "png": u"optipng" + exe + " -force -o7 '%(file)s'&&advpng" + exe + " -z4 '%(file)s' && pngcrush -rem gAMA -rem alla -rem cHRM -rem iCCP -rem sRGB -rem time '%(file)s' '%(file)s.bak' && mv '%(file)s.bak' '%(file)s'"
        }
        # Create a backup file
        copy(self.fullpath, self.fullpath + '~')
        try:
            retcode = call(runString[self.filetype] % {"file": self.fullpath},
                shell=True, stdout=PIPE)
        except:
            retcode = -1
        if retcode == 0:
            self.newfilesize = QFile(self.fullpath).size()
            self.compressed = True

            # Checks the new file and copy the backup
            if self.newfilesize >= self.oldfilesize:
                copy(self.fullpath + '~', self.fullpath)
                self.newfilesize = self.oldfilesize

            # Removes the backup file
            remove(self.fullpath + '~')
        else:
            self.failed = True
        self.compressing = False
        self.retcode = retcode
        return self


class Worker(QThread):

    def __init__(self, parent=None):
        QThread.__init__(self, parent)
        self.toDisplay = Queue()
        self.threadpool = ThreadPool(max_workers=cpu_count())

    def __del__(self):
        self.threadpool.shutdown()

    def compress_file(self, images, showapp, verbose, imagelist):
        """Start the worker thread."""
        for image in images:
            #FIXME:http://code.google.com/p/pythonthreadpool/issues/detail?id=5
            time.sleep(0.05)
            self.threadpool.add_job(image.compress, None,
                                    return_callback=self.toDisplay.put)
        self.showapp = showapp
        self.verbose = verbose
        self.imagelist = imagelist
        self.start()

    def run(self):
        """Compress the given file, get data from it and call update_table."""
        tp = self.threadpool
        while self.showapp or not (tp._ThreadPool__active_worker_count == 0 and
                                   tp._ThreadPool__jobs.empty()):
            image = self.toDisplay.get()

            self.emit(SIGNAL("updateUi"))

            if not self.showapp and self.verbose: # we work via the commandline
                if image.retcode == 0:
                    ir = ImageRow(image)
                    print("File: " + ir['fullpath'] + ", Old Size: "
                        + ir['oldfilesizestr'] + ", New Size: "
                        + ir['newfilesizestr'] + ", Ratio: " + ir['ratiostr'])
                else:
                    print >> sys.stderr, u"[error] %s could not be compressed" % image.fullpath


class Systray(QWidget):

    def __init__(self, parent):
        QWidget.__init__(self)
        self.parent = parent
        self.createActions()
        self.createTrayIcon()
        self.trayIcon.show()

    def createActions(self):
        self.quitAction = QAction(self.tr("&Quit"), self)
        QObject.connect(self.quitAction, SIGNAL("triggered()"),
            qApp, SLOT("quit()"))

        self.addFiles = QAction(self.tr("&Add and compress"), self)
        icon = QIcon()
        icon.addPixmap(QPixmap(self.parent.ui.get_image(("pixmaps/list-add.png"))),
            QIcon.Normal, QIcon.Off)
        self.addFiles.setIcon(icon)
        QObject.connect(self.addFiles, SIGNAL("triggered()"), self.parent.file_dialog)

        self.recompress = QAction(self.tr("&Recompress"), self)
        icon2 = QIcon()
        icon2.addPixmap(QPixmap(self.parent.ui.get_image(("pixmaps/view-refresh.png"))),
            QIcon.Normal, QIcon.Off)
        self.recompress.setIcon(icon2)
        self.recompress.setDisabled(True)
        QObject.connect(self.addFiles, SIGNAL("triggered()"), self.parent.recompress_files)

        self.hideMain = QAction(self.tr("&Hide window"), self)
        QObject.connect(self.hideMain, SIGNAL("triggered()"), self.parent.hide_main_window)

    def createTrayIcon(self):
        self.trayIconMenu = QMenu(self)
        self.trayIconMenu.addAction(self.addFiles)
        self.trayIconMenu.addAction(self.recompress)
        self.trayIconMenu.addSeparator()
        self.trayIconMenu.addAction(self.hideMain)
        self.trayIconMenu.addSeparator()
        self.trayIconMenu.addAction(self.quitAction)

        if QSystemTrayIcon.isSystemTrayAvailable():
            self.trayIcon = QSystemTrayIcon(self)
            self.trayIcon.setContextMenu(self.trayIconMenu)
            self.trayIcon.setToolTip("Optimgaika image compressor")
            self.trayIcon.setIcon(QIcon(self.parent.ui.get_image("pixmaps/optimgaika-icon.png")))


if __name__ == "__main__":
    app = QApplication(sys.argv)
    myapp = StartQT4()

    if myapp.showapp:
        myapp.show()
    sys.exit(app.exec_())
